diff --git a/xen/common/efi/boot.c b/xen/common/efi/boot.c
index d6c368fa09..0ababd8da2 100644
--- a/xen/common/efi/boot.c
+++ b/xen/common/efi/boot.c
@@ -1058,48 +1058,17 @@ static UINTN __init efi_find_gop_mode(EFI_GRAPHICS_OUTPUT_PROTOCOL *gop,
     return gop_mode;
 }
 
-static void __init efi_verify_kernel(EFI_HANDLE ImageHandle)
+/*
+ * Verifies the dom0 kernel signature via SHIM_LOCK->Verify
+ */
+static void __init efi_verify_kernel(EFI_SHIM_LOCK_PROTOCOL *shim_lock)
 {
-    static EFI_GUID __initdata shim_image_guid = SHIM_IMAGE_LOADER_GUID;
-    static EFI_GUID __initdata shim_lock_guid = SHIM_LOCK_PROTOCOL_GUID;
-    SHIM_IMAGE_LOADER *shim_loader;
-    EFI_HANDLE loaded_kernel;
-    EFI_SHIM_LOCK_PROTOCOL *shim_lock;
     EFI_STATUS status;
-    bool verified = true;
-    return;
 
-    /* Look for SHIM_IMAGE_LOADER first. */
-    if ( !EFI_ERROR(efi_bs->LocateProtocol(&shim_image_guid, NULL,
-                                           (void **)&shim_loader)) )
+    if ( efi_secureboot_enabled() )
     {
-        status = shim_loader->LoadImage(false, ImageHandle, NULL,
-                                        (void *)kernel.ptr, kernel.size,
-                                        &loaded_kernel);
-        if ( !EFI_ERROR(status) )
-            verified = true;
-
-        /*
-         * Always unload the image.  We only needed LoadImage() to perform
-         * verification anyway, and in the case of a failure there may still
-         * be cleanup needing to be performed.
-         */
-        shim_loader->UnloadImage(loaded_kernel);
+        if ( (status = shim_lock->Verify(kernel.ptr, kernel.size)) != EFI_SUCCESS )
+            PrintStr(L"Dom0 kernel image could not be verified");
     }
-
-    /* Otherwise, fall back to SHIM_LOCK. */
-    if ( !verified &&
-         !EFI_ERROR(efi_bs->LocateProtocol(&shim_lock_guid, NULL,
-                                           (void **)&shim_lock)) &&
-         shim_lock->Verify(kernel.ptr, kernel.size) == EFI_SUCCESS )
-        verified = true;
-
-    if ( !verified )
-    {
-        PrintStr(L"Kernel was not verified\n");
-
-        if ( efi_secure_boot )
-            blexit(L"Refusing to boot unverified kernel with UEFI SecureBoot enabled");
-    }
 }
 
@@ -1393,11 +1364,14 @@ void EFIAPI __init noreturn efi_start(EFI_HANDLE ImageHandle,
                                       EFI_SYSTEM_TABLE *SystemTable)
 {
     static EFI_GUID __initdata loaded_image_guid = LOADED_IMAGE_PROTOCOL;
+    static EFI_GUID __initdata shim_lock_guid = SHIM_LOCK_PROTOCOL_GUID;
+    static CHAR16 __initdata str_openxt[] = L"openxt.cfg";
     EFI_LOADED_IMAGE *loaded_image;
     EFI_STATUS status;
     unsigned int i;
-    CHAR16 *file_name, *cfg_file_name = NULL, *options = NULL;
+    CHAR16 *file_name, *cfg_file_name = str_openxt, *options = NULL;
     UINTN gop_mode = ~0;
+    EFI_SHIM_LOCK_PROTOCOL *shim_lock = NULL;
     EFI_GRAPHICS_OUTPUT_PROTOCOL *gop = NULL;
     union string section = { NULL }, name;
     bool base_video = false;
@@ -1485,6 +1459,10 @@ void EFIAPI __init noreturn efi_start(EFI_HANDLE ImageHandle,
     PrintStr(L"Xen " XEN_VERSION_STRING XEN_EXTRAVERSION
 	     " (c/s " XEN_CHANGESET ") EFI loader\r\n");
 
+    efi_bs->LocateProtocol(&shim_lock_guid, NULL, (void **)&shim_lock);
+    if ( !shim_lock )
+        PrintStr(L"No Shim Lock Protocol\r\n");
+
     efi_arch_relocate_image(0);
 
     if ( use_cfg_file )
@@ -1526,6 +1504,14 @@ void EFIAPI __init noreturn efi_start(EFI_HANDLE ImageHandle,
         }
         else if ( !read_file(dir_handle, cfg_file_name, &cfg, NULL) )
             blexit(L"Configuration file not found.");
+
+        if ( !shim_lock )
+        {
+            PrintStr(L"No shim found, booting xen.efi directly with OpenXT is not supported.");
+            efi_bs->Stall(10000000);
+            blexit(L"No shim found");
+        }
+
         pre_parse(&cfg);
 
         if ( section.w )
@@ -1648,8 +1634,8 @@ void EFIAPI __init noreturn efi_start(EFI_HANDLE ImageHandle,
      * device tree through the efi_check_dt_boot function, in this stage
      * verify it.
      */
-    if ( kernel.ptr && !kernel_verified )
-        efi_verify_kernel(ImageHandle);
+    if ( kernel.ptr && shim_lock )
+        efi_verify_kernel(shim_lock);
 
     efi_arch_edd();
 
diff --git a/xen/common/efi/runtime.c b/xen/common/efi/runtime.c
index 2385371914..b78e8da802 100644
--- a/xen/common/efi/runtime.c
+++ b/xen/common/efi/runtime.c
@@ -215,6 +215,34 @@ void efi_reset_system(bool warm)
     printk(XENLOG_WARNING "EFI: could not reset system (%#lx)\n", status);
 }
 
+bool efi_secureboot_enabled(void)
+{
+    EFI_GUID efi_gv = EFI_GLOBAL_VARIABLE;
+    static CHAR16 __initdata str_SecureBoot[] = L"SecureBoot";
+    static CHAR16 __initdata str_SetupMode[] = L"SetupMode";
+    UINT8 SecureBoot = 0, SetupMode = 0;
+    UINTN DataSize = sizeof(SecureBoot);
+    EFI_STATUS efi_status;
+
+    efi_status = efi_rs->GetVariable(str_SecureBoot, &efi_gv, NULL,
+                                     &DataSize, &SecureBoot);
+    if ( EFI_ERROR(efi_status) )
+        return false;
+
+    if ( !SecureBoot )
+        return false;
+
+    efi_status = efi_rs->GetVariable(str_SetupMode, &efi_gv, NULL,
+                                     &DataSize, &SetupMode);
+    if ( EFI_ERROR(efi_status) )
+        return false;
+
+    if ( SetupMode )
+        return false;
+
+    return true;
+}
+
 #endif /* CONFIG_ARM */
 
 const CHAR16 *wmemchr(const CHAR16 *s, CHAR16 c, UINTN n)
diff --git a/xen/include/xen/efi.h b/xen/include/xen/efi.h
index a08561ff45..8f071b8ad8 100644
--- a/xen/include/xen/efi.h
+++ b/xen/include/xen/efi.h
@@ -51,5 +51,6 @@ int efi_runtime_call(struct xenpf_efi_runtime_call *op);
 #endif
 int efi_compat_get_info(uint32_t idx, union compat_pf_efi_info *info);
 int efi_compat_runtime_call(struct compat_pf_efi_runtime_call *op);
+bool efi_secureboot_enabled(void);
 
 #endif /* __XEN_EFI_H__ */
